tools/swift-format/Sources/Rules/GroupNumericLiterals.swift

```diff
@@ -10,7 +10,67 @@ import SwiftSyntax
 ///
 /// Format: All numeric literals that should be grouped will have `_`s inserted where appropriate.
 ///
+/// TODO: Minimum numeric literal length bounds and numeric groupings selected arbitrarily, could
+///       be  reevaluated.
+///
+/// TODO: Handle floating point literals
+///
 /// - SeeAlso: https://google.github.io/swift#numeric-literals
 public final class GroupNumericLiterals: SyntaxFormatRule {
+  public override func visit(_ node: IntegerLiteralExprSyntax) -> ExprSyntax {
+    
+    var strDigits = node.digits.text
+    guard !strDigits.contains("_") else { return node }
+    
+    let isNegative = strDigits.first == "-"
+    strDigits = isNegative ? String(strDigits.dropFirst()) : strDigits
+    
+    var newDigits = ""
+    
+    switch strDigits.prefix(2) {
+    case "0x":
+      // Hexadecimal
+      let digitsNoPrefix = String(strDigits.dropFirst(2))
+      guard let intDigits = Int(digitsNoPrefix, radix: 16) else { return node }
+      guard intDigits >= 0x1000_0000 else { return node }
+      diagnose(.groupNumericLiteral(by: 4), on: node)
+      newDigits = "0x" + groupDigits(digitStr: digitsNoPrefix, by: 4)
+    case "0b":
+      // Binary
+      let digitsNoPrefix = String(strDigits.dropFirst(2))
+      guard let intDigits = Int(digitsNoPrefix, radix: 2) else { return node }
+      guard intDigits >= 0b1_000000000 else { return node }
+      diagnose(.groupNumericLiteral(by: 8), on: node)
+      newDigits = "0b" + groupDigits(digitStr: digitsNoPrefix, by: 8)
+    case "0o":
+      // Octal
+      return node
+    default:
+      // Decimal
+      guard let intDigits = Int(strDigits) else { return node }
+      guard intDigits >= 1_000_000 else { return node }
+      diagnose(.groupNumericLiteral(by: 3), on: node)
+      newDigits = groupDigits(digitStr: strDigits, by: 3)
+    }
+    
+    newDigits = isNegative ? "-" + newDigits : newDigits
+    return node.withDigits(SyntaxFactory.makeIdentifier(newDigits))
+  }
+  
+  func groupDigits(digitStr: String, by: Int) -> String {
+    var newGrouping = Array(digitStr)
+    var i = 1
```

Comment:
It looks like this algorithm works; a different way that might be a little easier to follow would be to start at `count`, then have a loop that subtracts `by` and inserts at that position until the index reaches 0. You won't have to worry about the indexes shifting like you do in the forward case because you'll always be inserting at lower positions than you did previously.